RWAHS Navigation plugin

leftclickben opened the pull request:
For RWAHS-253 - custom Find navigation menu items.
- Add a (configurable) list of custom links to a specific combination
  of search form and display (i.e. custom "Advanced Search" links).
- Basic configuration disables the plugin and specifies no custom links.

Configuration is at https://github.com/rwahs/providence-configuration/pull/18

Commit:
RWAHS Navigation plugin.

+ Add a (configurable) list of custom links to a specific combination
  of search form and display (i.e. custom "Advanced Search" links).
+ Basic configuration disables the plugin and specifies no custom links.

leftclickben commented on app/plugins/rwahsNavigation/conf/rwahsNavigation.conf:
```diff
@@ -0,0 +1,9 @@
+
+# Custom search shortcuts hash.  Keys are used as navigation keys; each value has a `label`, `form_id` and `display_id`.
+# Both form_id and display_id are numeric values, the database ID of the `search_form` or `display` record respectively.
+advanced_search_shortcuts = {
```
See https://github.com/rwahs/providence-configuration/pull/18 for the actual config.

leftclickben commented on app/plugins/rwahsNavigation/rwahsNavigationPlugin.php:
```diff
@@ -0,0 +1,55 @@
+    public function checkStatus() {
+        return array(
+            'description' => $this->getDescription(),
+            'errors' => array(),
```
TODO: check configuration structure

leftclickben commented on app/plugins/rwahsNavigation/rwahsNavigationPlugin.php:
```diff
@@ -0,0 +1,55 @@
+        $vo_custom_items = array();
+        $vo_spacer = array(
+            'spacer' => array(
+                'displayName' => '<div class="sf-spacer"></div>'
```
Apparently this is how spacers are done.

leftclickben commented on app/plugins/rwahsNavigation/rwahsNavigationPlugin.php:
```diff
@@ -0,0 +1,55 @@
+                'displayName' => $vo_shortcut['label'],
+                'default' => array(
+                    'module' => 'find',
+                    'controller' => 'SearchObjectsAdvanced',
```
Just objects because anything else YAGNI.

leftclickben commented on app/plugins/rwahsNavigation/rwahsNavigationPlugin.php:
```diff
@@ -0,0 +1,55 @@
+                ),
+                'parameters' => array(
+                    'form_id' => 'string:' . $vo_shortcut['form_id'],
+                    'display_id' => 'string:' . $vo_shortcut['display_id']
```
And this is how you pass a parameter with a "static" value.

leftclickben commented on app/plugins/rwahsNavigation/rwahsNavigationPlugin.php:
```diff
@@ -0,0 +1,55 @@
+                )
+            );
+        }
+        $pa_nav_info['find']['navigation'] = $vo_custom_items + $vo_spacer + $pa_nav_info['find']['navigation'];
```
Union operator ftw.

kehh commented:
As discussed today:
- [x] add type_id into the menu item
- [x] change configuration to use code/idno for display, form and item type
- [x] also add menu for `New` menu
- [x] Don't let `Museum Record` create page

Commit:
Use codes instead of ids for form and display.

Commit:
Update configuration documentation.

Commit:
Support optional `type_id` in navigation plugin config.

Commit:
Report on invalid configuration.

Commit:
Add check for presence of `label`.

Commit:
Add configuration check tests for RWAHS navigation plugin.

Commit:
WIP Integration tests for RWAHS navigation plugin.

Commit:
WIP Clean up documentation and remove accidentally `return`.

Commit:
Add missing attributes for search forms and displays.

Commit:
Fix typo in config file for test.

Commit:
Use correct call to get PK and add missing array elements in test.

Commit:
Support `options` in base test method.

+ Previously a `type code` could be passed, this was not
  used however is still supported via `options`.
+ Now `options` are passed instead as an array.
+ Also support custom labels via `options`.

Commit:
Complete `rwahsNavigationPlugin` functionality and tests.

+ Previously, only the `Find` menu was modified, now custom items
  are also added to the `New` menu based on configuration.
+ Each `New` menu item is only a `type_code`, which specifies the
  type to create and also the label, which is the name of the type.
+ The `Find` menu now also derives its labels from the type name,
  which means that `type_code` is now mandatory.
+ Also add two "hardcoded" items to the `Find` menu - basic search
  and browse; these are identical to the equivalent options in the
  "native" CA navigation menu.

leftclickben commented:
Also done:
- Add basic search and browse into top-level Find menu.
- Derive labels from `type_code` rather than hardcoding into configuration.
- Test coverage.

kehh commented on app/plugins/rwahsNavigation/rwahsNavigationPlugin.php:
```diff
@@ -0,0 +1,194 @@
+        if (is_array($va_shortcuts)) {
+            foreach ($va_shortcuts as $vs_key => $va_shortcut) {
+                if (!is_array($va_shortcut)) {
+                    array_push($va_errors, _t('Custom search shortcut with key "%1" is not an array', $vs_key));
```
I know it's just style, but what's wrong with `$va_errors[] =` instead of `array_push`? Have I been doing something wrong in the past?

leftclickben replied:
No nothing wrong, I just figured `array_push` fit better with `is_array` and `isset`.

kehh commented on app/plugins/rwahsNavigation/rwahsNavigationPlugin.php:
```diff
@@ -0,0 +1,194 @@
+        if (is_array($va_shortcuts)) {
+            foreach ($va_shortcuts as $vs_key => $vs_type_code) {
+                $vo_type = ca_list_items::find(
+                    array( 'idno' => $vs_type_code ),
```
Also add a restriction for `list_id` or `list_code` - `idno`'s are only meant to be unique within a specific list.

``` php
$t_subject = $this->opo_datamodel->getInstanceByTableName('ca_objects', true);
$t_list = new ca_lists();
$t_list->load(array('list_code' => $t_subject->getTypeListCode()));
$vn_list_id = $t_list->getPrimaryKey();
```

 Additionally we're probably going to have type based access control, so we'll probably need to check whether users can create objects of each type - as per https://github.com/rwahs/providence/blob/e2df4df82563e5412ff847600dddd0182db809d7/app/lib/ca/BaseEditorController.php#L1183-L1186
and later in that file.

leftclickben replied:
OK cool.  Type based access control will be separate, but I'll add the `list_code` here.

kehh commented on app/plugins/rwahsNavigation/rwahsNavigationPlugin.php:
```diff
@@ -0,0 +1,194 @@
+                    array( 'returnAs' => 'firstModelInstance' )
+                );
+                $vn_form_id = ca_search_forms::find(
+                    array( 'form_code' => $va_shortcut['form_code'] ),
```
add `table_num` for objects.

kehh replied:
``` php
$o_dm = Datamodel::load();
$vn_table_num = $o_dm->getTableNum('ca_objects');
```

leftclickben replied:
OK will do, but will try to use `getAppDatamodel` rather than `load()`ing it.

leftclickben replied:
Oops just read your other comment, so I won't add the table number here.

kehh commented on app/plugins/rwahsNavigation/rwahsNavigationPlugin.php:
```diff
@@ -0,0 +1,194 @@
+                    array( 'returnAs' => 'firstId' )
+                );
+                $vn_bundle_display_id = ca_bundle_displays::find(
+                    array( 'display_code' => $va_shortcut['display_code'] ),
```
add `table_num` here as well.

kehh replied:
Well I looked at the the schema, and `display_code` and `form_code` are both unique, so you don't need to add `table_num` in here or above (although that uniqueness is not restricted in `profile.xsd`).

kehh commented on tests/plugins/rwahsNavigation/RwahsNavigationPluginConfigurationTest.php:
```diff
@@ -0,0 +1,101 @@
+	public function testCorrectMissingNewMenuShortcutPropertiesGenerateErrors() {
+		$vo_plugin = new rwahsNavigationPlugin(__DIR__ . '/conf/invalid-new-menu-properties');
+		$va_plugin_status = $vo_plugin->checkStatus();
+		$this->assertEquals(2, sizeof($va_plugin_status['errors']), 'There is one error in the configuration file');
```
can `sizeof($va_plugin_status['errors'])` only equal 1 or 2? If not then the description of this error is incorrect.

leftclickben replied:
Copypasta, I believe.

kehh commented:
OK the main thing is enabling type based access control, but potentially that can be done under a task for that (I've just added https://youtrack.gaiaresources.com.au/youtrack/issue/RWAHS-321).

Commit:
Fix copypasta error in test.

Commit:
Restrict lookup of object types to the object type list.
